clippy_lints/src/vec_extend_to_append.rs

```diff
@@ -0,0 +1,53 @@
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use clippy_utils::diagnostics::span_lint_and_help;
+
+
+declare_clippy_lint! {
+    /// **What it does:** Finds occurrences of `vec.extend(other_vec.drain(..))`
+    ///
+    /// **Why is this bad?** Slightly more succinct code.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// vec![].extend(vec![1, 2, 3].drain(..));
+    /// ```
+    pub VEC_EXTEND_TO_APPEND,
+    perf,
+    "use append to a vector instead"
+}
+
+declare_lint_pass!(VecExtendToAppend => [VEC_EXTEND_TO_APPEND]);
+
+
+impl<'tcx> LateLintPass<'tcx> for VecExtendToAppend {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if is_vec_an_append(cx, expr) {
+            span_lint_and_help(
+                cx,
+                VEC_EXTEND_TO_APPEND,
+                expr.span,
+                "use of `vec.extend(other_vec.drain(..))`",
+                None,
+                "consider using `vec![].append(&mut vec![]);` instead",
+            );
+        }
+    }
+}
+
+fn is_vec_an_append<'tcx>(_cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>)->bool{
+    if_chain!{
+        if let ExprKind::MethodCall(method_name, _, exprs, _) = expr.kind;
+        if method_name.ident.as_str() == "extend";
+        if let ExprKind::MethodCall(method_name2, _, _exprs2, _) = &exprs[1].kind;
+        if method_name2.ident.as_str() == "drain";
```

Comment:
Your idea of this implementation is good, but there are some missing aspects.

1. You don't check for the specific types here. This means that any `a.extend(b.drain(..))` would be linted even if a and b are not of type `std::vec::Vec`
    * You can check the types by first getting the type of `exprs[0]` (the self object) and then using [`is_type_diagnostic_item`](https://github.com/rust-lang/rust-clippy/blob/a248648862cabc05a5fefe320bd9fcaa9a0a8c7b/clippy_utils/src/ty.rs#L234-L237)
    ```rs
    // Draft implementation
    let ty = cx.typeck_results().expr_ty(exprs[0]).peel_refs();
    is_type_diagnostic_item(cx, ty, sym::vec_type)
    ```
2. The implementation also doesn't check the given range inside the drain call. For instance `a.extend(b.drain(1..3))` should be valid as it's not equivalent to `a.append(&mut b)`
3. Now the next thing is probably something you couldn't know. Clippy already as a bunch of lints focusing on methods. They have therefor all been combined in a single module and lint pass. New lints focussing on methods should be injected into the existing context. The [`STRING_EXTEND_CHARS`]() is probably a good example for you as it does something very similar. Here are some links:
    * Injecting into current method lint-pass [-->](https://github.com/rust-lang/rust-clippy/blob/cd241b33cb8379311286f53c4afc6e671ddfded1/clippy_lints/src/methods/mod.rs#L1985)
    * Lint implementation: [-->](https://github.com/rust-lang/rust-clippy/blob/21083875d211c29fcfa4a21fcd66d4601d2b618b/clippy_lints/src/methods/string_extend_chars.rs#L13)
    * You can also notice that that implementation provides a suggestion with applicability. You can ignore that part for now and focus on the other notes.

Reply:
> 2\. a.extend(b.drain(1..3))

Thanks for the suggestion. Updated the lint and it seems to be working for the most part. Just one slight thing left. In the case of `a.extend(b.drain(1..3))` how do I get the `drain` func args from the exprs2 to verify that is its `1..3`? Thanks